### PR TITLE
Memory allocations

### LINKED ISSUE #39: Remove runtime memory allocation
Remove calls that trigger automatic malloc, and are of the type
```
Call routine(a, b(1:n)*c(1:n), ...)
```

  1. https://github.com/ralna/RALFit/blob/b9f438c25874379a71922f7c18f352722d7cf55c/libRALFit/src/ral_nlls_internal.f90#L380

   2. https://github.com/ralna/RALFit/blob/b9f438c25874379a71922f7c18f352722d7cf55c/libRALFit/src/ral_nlls_internal.f90#L2731

  3. https://github.com/ralna/RALFit/blob/9268bb743d36f56c7070a91226fb6e2aedf61a35/libRALFit/src/ral_nlls_internal.f90#L1549

### PR DESCRIPTION
**Pull request by tyronerees**
Fixes #39

### CODE REVIEW

**Comment by tyronerees** [comment]
@talassio : Apparently I did the automatic array work before Christmas, but forgot to ask for a review -- could you look over this and check I haven't missed anything.  I'll then merge it into `box-logic` (I did a test merge, and it's not too painful)

File: libRALFit/src/ral_nlls_internal.f90 [hunk 1]
```diff
@@ -295,6 +295,7 @@ recursive subroutine nlls_iterate(n, m, X,                   &
        if ( present(weights)) then
           ! set f -> Wf
           w%f(1:m) = weights(1:m)*w%f(1:m)
```
[the remark this answers is not in the document]
Comment (tyronerees) [reply libRALFit/src/ral_nlls_internal.f90]: I'm not sure I see why -- I think this repeats the functionality from before.  I'm probably missing something that you've seen, though.

[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_internal.f90]: Indeed, further inspection shows that it assumes `f` is weighted.

File: libRALFit/src/ral_nlls_internal.f90 [hunk 2]
```diff
@@ -295,6 +295,7 @@ recursive subroutine nlls_iterate(n, m, X,                   &
...
+          w%Wf(1:m) = weights(1:m)*w%f(1:m)
```
Comment (talassio) [inline libRALFit/src/ral_nlls_internal.f90]: Would it be convenient to just copy `w%f`, say `w%Wf(1:m) = w%f(1:m)`?

**Pull request merged**